feat: add additional metrics from pg tables

ErikOrm opened the pull request:
To improve visibility of the systems we tune, this adds a number of collectors from core pg-native statistics tables. 

#### Content
This adds a number of additional collectors. As many statistics are fetched from the same tables, keeping the "one collector per metric" would be highly inefficient, both in terms of runtime and code structure. I have instead made one collector per table that emits several metrics each. The `pg_stat_database`-collector has also take over the responsibility to calculate the `cache_hit_ratio`.

To lower the amount of duplicate code, I added two helper functions `EmitMetric` and `EmitDeltaMetricsMap`.
 
#### Backend compatibility
The data from pg_stat_user_tables comes in a per table format. For this, I added 3 new return types:
- map[str]int64
- map[str]float64
- map[str]time.Time

Those are currently not supported by the backend and will trigger a code: 400 error response. We can deal with that later, once we are satisfied with the agent changes.

#### PG-Version dependence
`pg_stat_checkpointer` only exists for PG 17 and higher. I ignore it for lower versions.
`pg_stat_wal` was introduced in PG 14, but PG 13 is EOL in November so I won't introduce a check for that yet.

For the checkpointer, the stats exist under different names before 17, but I haven't included that as it seems like more effort than it is worth at this point and would clutter the codebase. 

#### Testing
I have so far only tested running this on local on-prem pg, where it seems to  work fine.

But we have not tested any of the other Postgres flavours or other versions etc. 


#### Error handling
There is some error handling, but I might need to look at if there are any error cases I have missed. Especially, will grouping cached metrics into structs worsen fail safety?


#### More metrics
I have added the main metrics I could think of, but there might be more in the future.

eddiebergman commented on pkg/agent/agent.go:
```diff
@@ -175,6 +175,49 @@ type BufferStat struct {
+	BlksRead              int64
+	TuplesReturned        int64
+	TuplesFetched         int64
+	TuplesInserted        int64
+	TuplesUpdated         int64
+	TuplesDeleted         int64
+	TempFiles             int64
+	TempBytes             int64
+	Deadlocks             int64
+	IdleInTransactionTime float64
+	Timestamp             time.Time
+}
+
+type PGBGWriter struct {
+	BuffersClean    int64
+	MaxWrittenClean int64
+	BuffersAlloc    int64
+	Timestamp       time.Time
+}
+
+type PGWAL struct {
+	WALRecords     int64
+	WALFpi         int64
+	WALBytes       int64
+	WALBuffersFull int64
+	WALWrite       int64
+	WALSync        int64
+	WALWriteTime   float64
+	WALSyncTime    float64
+	Timestamp      time.Time
+}
+
+type PGCheckPointer struct {
+	NumTimed       int64
+	NumReq         int64
+	WriteTime      float64
+	SyncTime       float64
+	BuffersWritten int64
+	Timestamp      time.Time
+}
```
Not crucial but if you have doc links to where you got these from, it would be useful to include.

eddiebergman commented on pkg/aiven/adapter.go:
```diff
@@ -424,9 +430,24 @@ func AivenCollectors(adapter *AivenPostgreSQLAdapter) []agent.MetricCollector {
-			Key:        "database_cache_hit_ratio",
-			MetricType: "float",
-			Collector:  pg.BufferCacheHitRatio(pgDriver),
+			Key:        "pg_database",
+			MetricType: "int",
+			Collector:  pg.PGStatDatabase(pgDriver),
+		},
+		{
+			Key:        "pg_user_tables",
+			MetricType: "int",
+			Collector:  pg.PGStatUserTables(pgDriver),
+		},
+		{
+			Key:        "pg_bgwriter",
+			MetricType: "int",
+			Collector:  pg.PGStatBGwriter(pgDriver),
+		},
+		{
+			Key:        "pg_wal",
+			MetricType: "int",
+			Collector:  pg.PGStatWAL(pgDriver),
```
What happened to the `BufferCacheHitRatio`?

ErikOrm replied:
Moved to the database table as it only is based on the database metrics.

eddiebergman commented on pkg/aiven/adapter.go:
```diff
@@ -447,6 +468,14 @@ func AivenCollectors(adapter *AivenPostgreSQLAdapter) []agent.MetricCollector {
 			),
 		},
 	}
+	if adapter.PGVersion >= "17" {
```
We can't do naive string comparisons for versions as `"17.11" < "17.2"` by string comparison rules.

Also, `"9" > "17"`

eddiebergman commented on pkg/pg/collectors.go:
```diff
@@ -12,6 +13,47 @@ import (
+// Helper function to calculate the delta and emit a number of metrics
+// The mapping takes the current value, the previous value and the metric to emit the metric to
+// If either current or previous is nil, the metric is skipped
+// In some cases, the delta can be negative (e.g. if the counter has been reset)
```
Do we want to skip all metrics where the delta can be negative? This assumption is based on the fact that all metrics are mono-tonically increasing which I don't believe is true.

eddiebergman commented on pkg/pg/collectors.go:
```diff
@@ -12,6 +13,47 @@ import (
+	for _, m := range mappings {
+		if m.Previous != nil && m.Current != nil {
+			delta := *m.Current - *m.Previous
+			if *m.Previous >= 0 && *m.Current >= 0 && delta >= 0 {
```
Why the guard on `Previous` and `Current` being positive? Are all values we record strictly positive?

ErikOrm replied:
All cumulative metrics are strictly positive and monotonically increasing, yes.

eddiebergman commented on pkg/pg/collectors.go:
```diff
@@ -366,3 +327,426 @@ func WaitEvents(pgPool *pgxpool.Pool) func(ctx context.Context, state *agent.Met
+			return err
+		}
+
+		// Check if we have cached values from the previous collection
+		if state.Cache.PGDatabase.BlksHit > 0 && state.Cache.PGDatabase.BlksRead > 0 {
+			// Calculate deltas
+			blksHitDelta := blksHit - state.Cache.PGDatabase.BlksHit
+			blksReadDelta := blksRead - state.Cache.PGDatabase.BlksRead
+
+			// Handle counter resets (e.g., after PostgreSQL crash or immediate shutdown)
+			if blksHitDelta >= 0 && blksReadDelta >= 0 {
+				// Calculate cache hit ratio for this interval
+				var bufferCacheHitRatio float64
+				totalBlocks := blksHitDelta + blksReadDelta
+				if totalBlocks > 0 {
+					bufferCacheHitRatio = float64(blksHitDelta) / float64(totalBlocks) * 100.0
+				} else {
+					// No block activity in this interval
+					bufferCacheHitRatio = 0.0
+				}
+
+				// Validate the calculated ratio is within reasonable bounds
+				if bufferCacheHitRatio < 0.0 || bufferCacheHitRatio > 100.0 {
+					return fmt.Errorf("calculated cache hit ratio is out of bounds: %.2f%%", bufferCacheHitRatio)
+				}
+				CHRmetricEntry, err := metrics.PGCacheHitRatio.AsFlatValue(bufferCacheHitRatio)
+				if err != nil {
+					return err
+				}
+				state.AddMetric(CHRmetricEntry)
+			}
+		}
```
I'm not seeing where the `Previous` gets updated

eddiebergman commented on pkg/pg/collectors.go:
```diff
@@ -366,3 +327,426 @@ func WaitEvents(pgPool *pgxpool.Pool) func(ctx context.Context, state *agent.Met
+// PGStatUserTables collects statistics from pg_stat_user_tables, computes deltas for various metrics per table,
+// and emits them as metrics for monitoring purposes. It tracks metrics such as autovacuum counts, live/dead tuples,
+// and sequential/index scans, providing insights into table activity and performance over time.
+func PGStatUserTables(pgPool *pgxpool.Pool) func(ctx context.Context, state *agent.MetricsState) error {
+	return func(ctx context.Context, state *agent.MetricsState) error {
+		var jsonResult string
+		err := pgPool.QueryRow(ctx, PGStatUserTablesQuery).Scan(&jsonResult)
+		if err != nil {
+			if err.Error() == "can't scan into dest[0]: cannot scan NULL into *string" {
+				return errors.New("pg_stat_user_tables returned no data, likely this means there are no user tables in the database")
+			}
+			return err
+		}
+
+		var tableStats map[string]utils.PGUserTables
+		err = json.Unmarshal([]byte(jsonResult), &tableStats)
+		if err != nil {
+			return err
+		}
```
@mbocevski not for now but this is also likely something we want to be able to update from the frontend, i.e. the user can select which tables we are allowed to query.

eddiebergman commented on pkg/pg/collectors.go:
```diff
@@ -366,3 +327,426 @@ func WaitEvents(pgPool *pgxpool.Pool) func(ctx context.Context, state *agent.Met
+)
+as stats
+FROM pg_stat_user_tables
+as f
```
`as f`?

eddiebergman reviewed (commented):
Just a review of the draft. Main concerns:
1) cache hit ratio still works? I see it's moved in PGDatabase, but it's treated differently and the logic around it seems different.
2) Everything is treated as if it should have positive deltas, implying each metric is a monotonically increasing metric. I don't believe this is the case.

ErikOrm commented:
> Just a review of the draft. Main concerns:
> 
> 1. cache hit ratio still works? I see it's moved in PGDatabase, but it's treated differently and the logic around it seems different.
> 2. Everything is treated as if it should have positive deltas, implying each metric is a monotonically increasing metric. I don't believe this is the case.
> 3. We might need to be a bit more selective of the tables we get stats of and know about, i.e. there might be cases where someone doesn't want us storing that there's a "payments" table, along with its statistics.

Thanks for the review!

1. It was moved so that we don't have to query the table twice. I tried to clean it up, because it was messy, but the behaviour should be the same. They did the "save to cache" first, whereas I do the "save to cache" last, which makes the code much more readable.

2. All statistics I use are taken from the Postgres cumulative tables (https://www.postgresql.org/docs/current/monitoring-stats.html). No all of them are actually cumulative, but almost every statistic I have added are strictly positive and monotonically increasing. 

I have added two helper functions:
- EmitMetric
- EmitCumulativeMetricMap

The first of which I think should be general, but the second does indeed assume those properties. If those are not satisfied, it would be the wrong method to use. If there would be some code structure we could use to make it more clear which metrics are cumulative and which are not, that would be a great addition.  I will think about it.

3. Good point. This would require plenty of additional work to get user input involved. Until then I can remove the table names and work purely with IDs if we think that is better?

eddiebergman reviewed (approved):
Approve for changes since last review!
